File: cpp/Problem1995.cpp
/*
 * Problem: 1995. Count Special Quadruplets
 * URL: https://leetcode.com/problems/count-special-quadruplets/
 * Difficulty: Easy
 * Tags: Array
 */

#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
	int countQuadruplets(vector<int>& nums) {
		int result = 0;
		int size = nums.size();
		unordered_map<int, int> count;
		count[nums[size - 1] - nums[size - 2]] = 1;

		for (int i = size - 3; i > 0; i--) {
			for (int j = i - 1; j >= 0; j--) {
				result += count[nums[i] + nums[j]];
			}
			for (int k = size - 1; k > i; k--) {
				count[nums[k] - nums[i]]++;
			}
		}
		return result;
	}
};

File: python/Problem1995.py
# Problem: 1995. Count Special Quadruplets
# URL: https://leetcode.com/problems/count-special-quadruplets/
# Difficulty: Easy
# Tags: Array

from typing import List
from collections import defaultdict

class Solution:
  def countQuadruplets(self, nums: List[int]) -> int:
    result: int = 0
    length: int = len(nums)

    count: dict = defaultdict(lambda: 0)
    count[nums[length - 1] - nums[length - 2]] = 1

    for i in range(length - 3, 0, -1):
      for j in range(i - 1, -1, -1):
        result += count[nums[i] + nums[j]]
      for k in range(length - 1, i, -1):
        count[nums[k] - nums[i]] += 1
    return result